src/Bicep.Core/Semantics/Metadata/ResourceMetadata.cs

```diff
@@ -1,22 +1,54 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
+using System;
+using System.Linq;
+using Bicep.Core.Extensions;
 using Bicep.Core.Resources;
 using Bicep.Core.Syntax;
 using Bicep.Core.TypeSystem;
 using Bicep.Core.Semantics.Namespaces;
+using Bicep.Core.TypeSystem.Az;
+using System.Collections.Generic;
+using System.Collections.Immutable;
 
 namespace Bicep.Core.Semantics.Metadata
 {
     public record ResourceMetadata(
         ResourceType Type,
-        SyntaxBase NameSyntax,
         ResourceSymbol Symbol,
         ResourceMetadataParent? Parent,
-        SyntaxBase? ScopeSyntax,
         bool IsExistingResource)
     {
+        private readonly ImmutableDictionary<string, SyntaxBase> UniqueIdentifiers = GetUniqueIdentifiers(Type, Symbol);
+
         public ResourceTypeReference TypeReference => Type.TypeReference;
 
+        public SyntaxBase NameSyntax => UniqueIdentifiers[AzResourceTypeProvider.ResourceNamePropertyName];
```

Comment:
>UniqueIdentifiers [](http://example.com/codeflow?start=40&length=17)

This will throw a misleading exception when called incorrectly. Can you do `TryGetNameSyntax() ?? throw new ...` with a friendlier exception message?